_posts/2021-12-14-announcing-readonly.markdown

```diff
@@ -0,0 +1,94 @@
+---
+title: "Introducing `readonly`"
+layout: post
+author: jjwu
+category: blog
+---
+
+As of HHVM version 4.139, we've released a new keyword called `readonly` in Hack. `readonly` is a new feature in Hack for restricting the mutability of object properties. It's designed for use cases that require performant, runtime-enforced mutability control. 
+
+`readonly` has two main principles: **readonlyness* and **deepness*. 
+
+**Readonlyness**: Object properties of any readonly value cannot be modified (i.e. mutated).
+**Deepness**: All nested properties of a readonly value are readonly.
+
+```Hack
+class Bar {
+  public function __construct(
+    public Foo $foo,
+  ){}
+}
+class Foo {
+  public function __construct(
+    public int $prop,
+  ) {}
+}
+
+function test(readonly Foo $foo, readonly Bar $bar) : void { 
+  $foo->prop = 5; // error, $foo is readonly 
+
+  // Deepness: $x is also readonly
+  $x = $bar->foo;
+
+  $x->prop = 3; // error, $x is readonly
+}
+```
+
+The readonly keyword can be used in many places in Hack. 
+- **Readonly expressions**: Any expression can be marked readonly with the `readonly` keyword. 
+- **Readonly parameters**: A parameter marked readonly signal that a function or method can take in a readonly value, and does not modify it. 
+- **Readonly return types**: A function can return readonly, which allows it to return a readonly reference. 
+- **Readonly properties**: A property can be marked readonly, so that you can assign a readonly value to it. 
```

Comment:
So, a readonly property can have its' value replaced, as long as its' new value is itself readonly?

Perhaps make this more explicit?